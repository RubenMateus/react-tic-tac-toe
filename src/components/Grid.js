import React from 'react';
import { css } from '@emotion/css';
import PropTypes from 'prop-types';

import Cell from './Cell';

const Grid = ({ grid, onClick }) => {
  return (
    <div
      data-testid="grid"
      className={css`
        display: inline-block;
      `}
    >
      <div
        className={css`
          background-color: #444;
          display: grid;
          grid-template-rows: repeat(${grid.length}, 1fr);
          grid-template-columns: repeat(${grid[0].length}, 1fr);
          grid-gap: 2px;
        `}
      >
        {grid.map((row, rowIndex) =>
          row.map((column, colIndex) => (
            <Cell
              key={`${colIndex}-${rowIndex}`}
              value={column}
              onClick={() => onClick(colIndex, rowIndex)}
            />
          ))
        )}
      </div>
    </div>
  );
};

Grid.propTypes = {
  grid: PropTypes.array.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Grid;
