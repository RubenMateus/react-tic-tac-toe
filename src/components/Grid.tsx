import { css } from "@emotion/css";

import Cell from "./Cell";
import { State } from "../state/reducer";

type GridProps = {
  grid: State["grid"];
  onClick: (x: number, y: number) => void;
};

const Grid = ({ grid, onClick }: GridProps) => {
  return (
    <div
      data-testid="grid"
      className={css`
        display: inline-block;
      `}
    >
      <div
        className={css`
          background-color: #444;
          display: grid;
          grid-template-rows: repeat(${grid.length}, 1fr);
          grid-template-columns: repeat(${grid[0].length}, 1fr);
          grid-gap: 2px;
        `}
      >
        {grid.map((row, rowIndex) =>
          row.map((column, colIndex) => (
            <Cell
              key={`${colIndex}-${rowIndex}`}
              value={column}
              onClick={() => onClick(colIndex, rowIndex)}
            />
          ))
        )}
      </div>
    </div>
  );
};

export default Grid;
