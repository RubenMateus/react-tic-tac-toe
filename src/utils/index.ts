export const clone = (x: unknown) => JSON.parse(JSON.stringify(x));

export const generateGrid = (
  rows: number,
  columns: number,
  mapper: () => null | "X" | "O"
) =>
  Array(rows)
    .fill(null)
    .map(() => Array(columns).fill(null).map(mapper));
