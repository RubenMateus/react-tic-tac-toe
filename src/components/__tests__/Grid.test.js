import React from 'react';
import { render } from '@testing-library/react';
import Grid from '../Grid';
import { generateGrid } from '../../utils';

const emptyGrid = generateGrid(3, 3, () => null);

describe('<Grid />', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <Grid grid={emptyGrid} onClick={jest.fn()} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
