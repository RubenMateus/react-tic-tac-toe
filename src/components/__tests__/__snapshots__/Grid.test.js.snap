// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1w73uij"
  >
    <div
      class="css-1x1afi9"
    >
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
      <div
        class="css-ppfkvi"
      >
        <button
          class="css-113hvi"
          type="button"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
